pkg/cmd/roachtest-selector/sfselector/selector.go

```diff
@@ -0,0 +1,298 @@
+	}
+	// skipped tests are maintained separately
+	// this helps in considering them for running based on selectFromSuccessPct
+	skippedTests := make([][]string, 0)
+	td := &testDetails{
+		columnHeaders: colHeaders,
+		testInfos:     make([][]string, 0),
+	}
+	selectedTestCount := 0 // used for printing the number of tests selected
+	for rows.Next() {
+		_ = rows.Scan(colPointers...)
+		testInfos := make([]string, len(colContainer))
+		copy(testInfos, colContainer)
+		if testInfos[1] == "no" {
+			// skipped based on query
+			skippedTests = append(skippedTests, testInfos)
+		} else {
+			// selected for running
+			td.testInfos = append(td.testInfos, testInfos)
+			selectedTestCount++
+		}
+	}
+	if req.SelectFromSuccessPct > 0 && len(skippedTests) > 0 {
+		// need to select some tests from th success list (skipped tests)
+		numberOfTestsToSelect := len(skippedTests) * req.SelectFromSuccessPct / 100
+		// the tests are sorted by the number of runs. So, simply iterate over teh list
+		// and select the first count of "numberOfTestsToSelect"
+		for i := range skippedTests {
+			if i == numberOfTestsToSelect {
+				break
+			}
+			// mark the skipped tests as selected="yes"
+			skippedTests[i][1] = "yes"
+			selectedTestCount++
+		}
+	}
+	// add all the skipped tests (some may have been selected based on selectFromSuccessPct)
+	td.testInfos = append(td.testInfos, skippedTests...)
+	fmt.Printf("%d out of %d tests selected!\n", selectedTestCount, len(td.testInfos))
+	return td, nil
+}
+
+// getConnect makes connection to snowflake and returns the connection.
+func getConnect(ctx context.Context) (*gosql.DB, error) {
+	username, password, err := getSFCreds(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dsn, err := sf.DSN(&sf.Config{
+		Account:   account,
+		Database:  database,
+		Schema:    schema,
+		Warehouse: warehouse,
+		Password:  password,
+		User:      username,
+	})
+	if err != nil {
+		return nil, err
+	}
+	db, err := gosql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+// getCSVInMemory creates a CSV file in memory from a testDetails
+func getCSVInMemory(td *testDetails) (*bytes.Buffer, error) {
+	buffer := &bytes.Buffer{}
+	w := csv.NewWriter(buffer)
+	defer w.Flush()
+	if err := w.Write(td.columnHeaders); err != nil {
+		return nil, err
+	}
+	if err := w.WriteAll(td.testInfos); err != nil {
+		return nil, err
+	}
+	return buffer, nil
+}
+
+// uploadCSVToGCS uploads a CSV file in memory to Google Cloud Storage
+func uploadCSVToGCS(csvBuffer *bytes.Buffer, cloud, suite string) error {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, option.WithScopes(storage.ScopeFullControl), option.WithQuotaProject(project))
+	if err != nil {
+		return fmt.Errorf("failed to create GCS client: %w", err)
+	}
+	defer func() { _ = client.Close() }()
+
+	wc := client.Bucket(bucket).Object(
+		fmt.Sprintf("%s-%s-%s.%s", testsFileLocation, suite, cloud, testsCsvExtension)).NewWriter(ctx)
```

Comment:
We need to scope this to a given branch names as well, right? We might only be interested in enabling this on master today, but we likely will want this functionality to exist in release branches as well.